### ISSUE TITLE
make -fcompiler_rt default to enabled when building static libraries

### ISSUE THREAD

**Issue by markfirmware** [ziglang/zig#1993]
I also need this for armv6, and another test case was missing memcpy.
I see something similar in #254 which is closed.

debian9/x64
Linux localhost 4.9.0-8-amd64 #1 SMP Debian 4.9.130-2 (2018-10-27) x86_64 GNU/Linux    

    $ zig version
    0.3.0+db31c252

    $ cat file.zig
    const std = @import("std");

    export fn zigmain() void {
        var warnBuf: [1024]u8 = undefined;
        _ = std.fmt.bufPrint(&warnBuf, "");
    }

    $ zig build-obj --target-arch armv7 --target-os freestanding --target-environ gnueabihf file.zig
    $ nm file.o | grep U
            U __aeabi_uidivmod
            U memset

**Comment by andrewrk** [comment]
These functions are provided automatically by Zig when you use `build-exe`, but not `build-obj`.

One thing it may make sense to do is to provide these symbols but with `weak` linkage, for `build-lib --static`. So then instead of `zig build-obj --target-arch armv7 --target-os freestanding --target-environ gnueabihf file.zig` you would do `zig build-lib --static --target-arch armv7 --target-os freestanding --target-environ gnueabihf file.zig`. Does that sound good?

**Comment by markfirmware** [comment]
That sounds great. This is not holding me up, I just noticed that I was linking in an extra dependency that I don't want. Mark.

**Comment by andrewrk** [comment]
I want to note - LLVM emits libcalls to these functions; its generated code depends on them being available. Furthermore, when the optimizer runs it may collapse a for loop into a call to memset or memcpy, or other functions such as these. Zig provides all these functions in compiler_rt and builtin.

So you cannot escape them as dependencies, but you can rely on the dependencies always being fulfilled.

**Comment by markfirmware** [comment]
Oh - so they won't be supplied by zig's stdlib?

**Comment by andrewrk** [comment]
That's correct, they are supplied by zig's stdlib, in `std/special/builtin.zig` and `std/special/compiler_rt/*`. However it's common for these symbols to be supplied by external libs such as libc, libgcc, compiler_rt, and they have well-defined ABIs. So Zig supplies these dependencies in a way that will not take up double-space if they are ultimately fulfilled externally. For example, if you link libc, then Zig's implementation of `memcpy` and `memset` are garbage-collected by the linker thus waste no space.

**Comment by markfirmware** [comment]
Thanks for the attention. Great service!

**Comment by andrewrk** [comment]
Zig has `--bundle-compiler-rt` and corresponding build.zig API, so this is a low priority issue.

**Comment by markfirmware** [comment]
@andrewrk Thanks I will look at --bundle-compiler-rt. This issue arose when trying to integrate zig with https://ultibo.org which is taking lower priority to zig-only solutions.

**Comment by andrewrk** [comment]
I believe this use case is fully solved. Feel free to describe any related problem by opening a fresh issue on Codeberg.
